src/sentry/testutils/thread_leaks/diff.py

```diff
@@ -0,0 +1,109 @@
+"""Generate readable diffs showing thread creation locations for debugging leaks."""
+
+import os
+import sys
+from collections.abc import Generator, Iterable
+from threading import Thread
+from traceback import FrameSummary, StackSummary
+
+
+def get_relevant_frames(stack: Iterable[FrameSummary]) -> StackSummary:
+    """Filter stack frames to show only application code relevant for debugging.
+
+    Applies a series of filters to remove infrastructure code, keeping only
+    frames that help developers identify where threads were created. Each filter
+    is only applied if it doesn't remove all frames (fail-safe behavior).
+
+    Returns the last 10 frames of application code with relative paths.
+    """
+    # Get current working directory for path relativization
+    cwd = os.getcwd() + "/"
+
+    # Get stdlib directory path if available (for filtering system code)
+    stdlib_dir = getattr(sys, "_stdlib_dir", None)
+
+    # Apply each filter independently, reverting if it would remove all frames
+    for filter in (
+        lambda frame: frame.filename == __file__,  # Remove this very file
+        lambda frame: frame.line is None,  # Remove "frozen" stdlib modules
+        lambda frame: stdlib_dir
+        and frame.filename.startswith(stdlib_dir + "/"),  # Remove Python stdlib
+        lambda frame: frame.filename.startswith(sys.prefix),  # Remove system Python files
+        lambda frame: "/test_" not in frame.filename,  # Keep test files out of stack traces
+        lambda frame: "/testutils/" not in frame.filename,  # Remove testing infrastructure
+    ):
+        filtered_stack = [frame for frame in stack if not filter(frame)]
+        # Only apply the filter if it leaves some frames (fail-safe)
+        if filtered_stack:
+            stack = StackSummary.from_list(filtered_stack)
+
+    # Convert absolute paths to relative and limit to most recent frames
+    filtered_stack = [
+        FrameSummary(frame.filename.replace(cwd, "./"), frame.lineno, frame.name) for frame in stack
+    ]
+    # Return last 10 frames to keep output manageable while showing call chain
+    return StackSummary.from_list(filtered_stack[-10:])
```

Comment:
What does this look like in Sentry for over 10 items? I believe the product does something to display stacktraces nicely even if there are many frames involved

Reply:
I've removed the shortening.